source/manuals/programming-languages/python/python.html.md.erb

```diff
@@ -1,106 +1,150 @@
+as a complete web application, perhaps with many dependencies and sub-dependencies. It also
+applies to a collection of scripts that are deployed into the cloud and run automatically
+(for example, batch jobs).
 
-    **Recommendation**: Put your top-level requirements in a
-    `requirements-app.txt` file and use [this Makefile
-    script][dm-deps-commit-makefile] to generate a fully specified
-    `requirements.txt`, which is used whenever you need to pip-install the
-    dependencies. The script creates a temporary virtualenv, pip-installs the
-    `requirements-app.txt` and pip-freezes the result as `requirements.txt`.
+A good strategy for specifying your application's Python dependencies has two desirable
+characteristics - they should be:
 
-    **Recommendation**: Use `requirements-dev.txt` for any dependencies used in
-    development/test but not production environments.
+1. Reproducible (predictable)
 
+    Pin your application's full dependencies – specific versions, rather than ranges – or you'll
+    get unpredictability between your dev environment and other environments. You want a
+    new starter to avoid small hard-to-spot problems. And you want parity between what you
+    test locally, what is tested by CI, and what you deploy, or you risk new issues appearing on
+    a live server. Additionally these things can be hard to diagnose.
 
 2. Kept up-to-date
 
-    Security issues are found in libraries, so it is important to choose
-    libraries that are maintained and to ensure your team has a strategy to
-    ensure security updates are installed without significant delay.
+    Security issues are found in libraries, so it is important to choose libraries that are
+    maintained and to ensure your team has a strategy to ensure security updates are installed
+    without significant delay. The [how to manage third party software dependencies][gds-way-deps] section
+    gives further context and discusses tools that can help, such as [snyk.io][].
+
+Your README should document an easy-to-follow process by which all your Python
+dependencies can be upgraded to get bug fixes and security fixes, without introducing
+breaking changes to your build.
+
+Your pinned dependencies should be fully specified in a file called `requirements.txt`, and checked
+into your version control system (VCS). For projects with only a small number of dependencies, maintaining
+this manually (for example, installing with `pip install`, then using `pip freeze`) may be adequate.
+
+For larger projects, GDS recommends the following approach, which was developed after various issues
+were found in current mainstream tooling (see [this commit message][dm-deps-commit], and note also the
+[lack of support for specifying VCS dependencies](https://gds.slack.com/archives/CAD6NP598/p1541777461082500)
+in `pip-compile`).
+
+Put your top-level requirements in a `requirements-app.txt` file.
```

Comment:
What's meant by a `top-level requirement`? And whats the reasoning behind this?

Reply:
Top-level, as in: the libraries that this application depends upon, rather than the dependencies of _those_ dependencies (i.e. not the sub-dependencies).

They need to be specified somewhere. One option is to have them in a `setup.py` file (which is what we recommend for libraries); however, wrangling them is much easier when they are in plain text files. In `pip-compile`, the file would be called `requirements.in` and maybe we should consider using that name instead, or making clear that we really don't care what the name is, just that you follow this model for having top-level requirements and then a resulting pinned full dependency list.

If there's a form of words that could make this clear (assuming it now is) then please do suggest, very happy to modify this further.

Reply:
I think this is related to the question around `last responsible moment`; clearly what is needed here is an exposition of the dependency hierarchy in the introduction that defines the terms `dependency`, `sub-dependency`, and maybe clearly explains why the distinction is important when dealing with Python packaging.

Reply:
I've changed this to use the term 'immediate dependencies' which is defined in the new introductory text